swupd-extract/state.go

```diff
@@ -0,0 +1,367 @@
+	return filepath.Join(cs.dir, filepath.Join(elem...))
+}
+
+func (cs *clientState) GetBundleManifest(version, name, expectedHash string) (*swupd.Manifest, error) {
+	if name == "MoM" {
+		return nil, fmt.Errorf("invalid arguments to GetBundleManifest: MoM is not a bundle")
+	}
+	filename, err := cs.GetFile(version, "Manifest."+name)
+	if err != nil {
+		return nil, err
+	}
+	hash, err := swupd.GetHashForFile(filename)
+	if err != nil {
+		_ = os.Rename(filename, filename+".invalid")
+		return nil, fmt.Errorf("couldn't calculate hash for %s: %s", filename, err)
+	}
+	if hash != expectedHash {
+		_ = os.Rename(filename, filename+".invalid")
+		return nil, fmt.Errorf("hash mismatch in %s got %s but expected %s", filename, hash, expectedHash)
+	}
+	m, err := swupd.ParseManifestFile(filename)
+	if err != nil {
+		_ = os.Rename(filename, filename+".invalid")
+		return nil, err
+	}
+	return m, nil
+}
+
+func (cs *clientState) GetFullfile(version, hash string) error {
+	tarredFilename, err := cs.GetFile(version, "files", hash+".tar")
+	if err != nil {
+		return err
+	}
+	tarred, err := os.Open(tarredFilename)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = tarred.Close()
+	}()
+
+	tr, err := newCompressedTarReader(tarred)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = tr.Close()
+	}()
+
+	hdr, err := tr.Next()
+	if err != nil {
+		return err
+	}
+	err = cs.extractFullfile(hdr, tr)
+	if err != nil {
+		return err
+	}
+
+	hdr, err = tr.Next()
+	if err == nil {
+		fmt.Printf("! ignoring unexpected extra content in %s: %s\n", tarredFilename, hdr.Name)
+	}
+
+	return nil
+}
+
+func (cs *clientState) GetZeroPack(version, name string) error {
+	cachedName := cs.Path(fmt.Sprintf("pack-%s-from-0-to-%s.tar", name, version))
+	if !cs.noCache {
+		if _, err := os.Stat(cachedName); err == nil {
+			return nil
+		}
+	}
+
+	pack, err := cs.OpenFile(version, fmt.Sprintf("pack-%s-from-0.tar", name))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = pack.Close()
+	}()
+
+	tr, err := newTarXzReader(pack)
+	if err != nil {
+		return err
+	}
+	for {
+		var hdr *tar.Header
+		hdr, err = tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("invalid pack for %s", name)
+		}
+		if !strings.HasPrefix(hdr.Name, "staged/") || hdr.Name == "staged/" {
+			continue
+		}
+		err = cs.extractFullfile(hdr, tr)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = tr.Close()
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(cachedName, nil, 0600)
+}
+
+func download(u, path string) (err error) {
+	tempPath := path + ".downloading"
+	f, err := os.OpenFile(tempPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("couldn't open temporary file to write downloaded contents: %s", err)
+	}
+	defer func() {
+		if err != nil {
+			_ = os.Remove(tempPath)
+		}
+	}()
+	fmt.Printf("- downloading %s\n", u)
+	res, err := http.Get(u)
+	if err != nil {
+		_ = f.Close()
```

Comment:
Message if http.Get returns an error? You give one if the res.StatusCode is not 200...

Reply:
The Get error already contains the URL and the reason of failure.